rir/src/compiler/opt/eager_calls.cpp

```diff
@@ -14,91 +17,213 @@ namespace rir {
+
+        return ip;
+    };
+
+    auto replaceCallWithCallBuiltin = [&](BB* bb, BB::Instrs::iterator ip,
+                                          Call* call, SEXP builtin) {
+        std::vector<Value*> args;
+        call->eachCallArg([&](Value* a) {
+            args.push_back(a);
+            if (auto mk = MkArg::Cast(a))
+                if (!mk->isEager())
+                    todo.insert(mk);
+        });
+        auto bt =
+            BuiltinCallFactory::New(call->env(), builtin, args, call->srcIdx);
+        call->replaceUsesWith(bt);
+        bb->replace(ip, bt);
+    };
+
+    // Search for calls that likely point to a builtin.
+    std::unordered_map<LdFun*, SEXP> replaced;
     Visitor::run(code, [&](BB* bb) {
-        for (auto ip = bb->begin(); ip != bb->end(); ++ip) {
-            auto call = StaticCall::Cast(*ip);
-            if (!call)
-                continue;
-
-            Closure* cls = call->cls();
-            ClosureVersion* version = call->tryDispatch();
-
-            if (!version ||
-                call->nCallArgs() != cls->nargs())
-                continue;
-
-            bool allEager =
-                version->properties.includes(ClosureVersion::Property::IsEager);
-            if (!allEager && version->properties.argumentForceOrder.empty())
-                continue;
-
-            auto isEager = [&](size_t i) {
-                if (allEager)
-                    return true;
-                for (auto a : version->properties.argumentForceOrder) {
-                    // We know that a is forced before i, therefore we are not
-                    // in left-to-right order
-                    // TODO: support reordering of the evaluation
-                    if (a > i)
-                        return false;
-                    // We found the argument in the list of certainly forced
-                    // promises
-                    if (a == i)
-                        return true;
+        auto ip = bb->begin();
+        while (ip != bb->end()) {
+            auto next = ip + 1;
+
+            if (auto call = Call::Cast(*ip)) {
+                if (auto ldfun = LdFun::Cast(call->cls())) {
+                    if (ldfun->hint) {
+                        if (TYPEOF(ldfun->hint) == BUILTINSXP) {
+                            // We can only speculate if we have a checkpoint at
+                            // the ldfun position, since we want to deopt before
+                            // forcing arguments.
+                            if (checkpoint.at(ldfun)) {
+                                replaced.emplace(ldfun, ldfun->hint);
+                                replaceCallWithCallBuiltin(bb, ip, call,
+                                                           replaced.at(ldfun));
+                            }
+                        }
+                    } else {
+                        // Only speculate if we don't have a static guess
+                        if (!ldfun->guessedBinding()) {
```

Comment:
Maybe we can add profiling also for standard calls?

Reply:
sorry, I am not sure if I understand this comment. What do you mean by standard call. We should have profiling for all calls. But I guess sometimes it does not work....

Reply:
Oh yes, but then this seems to open the possibility to speculate even if we already observed during the warmup multiple targets.  Oth, I do not fully get the `Rf_findVar` here in a pir pass.

Reply:
> Oh yes, but then this seems to open the possibility to speculate even if we already observed during the warmup multiple targets

yes, actually your comment made me realize that and I fixed this by adding a dummy value in this case.

> Oth, I do not fully get the Rf_findVar here in a pir pass.

well, if we already statically see that the function is overwritten, then we don't need to speculate, since we know it will fail...